### PR TITLE
Add link repost tracker

### PR DESCRIPTION
**Pull request by maxgoedjen**
Adds a feature to Camille to let people know if a link is already being discussed in a different channel.

image

### CODE REVIEW

File: Sources/SlackBotKit/Camillink/Camillink+LinkMention.swift [hunk 1]
```diff
@@ -59,7 +60,10 @@ extension CamillinkService {
     
     func shouldSilenceForCrossLink(message: MessageDecorator, record: Record) -> Bool {
         guard config.silentCrossLink else { return false }
-        return message.mentionedChannels.map({ $0.value.id }).contains(record.channelID)
+        let ids = message.mentionedChannels.map({ $0.value.id })
```
Comment (bugKrusha) [inline Sources/SlackBotKit/Camillink/Camillink+LinkMention.swift]: The formatting here is a little off. Please update to:
```
return message
    .mentionedChannels
    .map { $0.value.id }
    .compactMap { $0.split(separator: "|").first }
    .map { String($0) }
    .contains(record.channelID)
```

File: Sources/SlackBotKit/Camillink/Camillink+LinkMention.swift [hunk 2]
```diff
@@ -17,7 +17,8 @@ extension CamillinkService {
     }
     
     func previousLinkDiscussion(linkURL: URL) throws -> Record? {
-        let string = try storage.get(key: linkURL.absoluteString, from: Keys.namespace, or: "INVALID")
+        let string = try storage.get(key: linkURL.absoluteString, from: Keys.namespace, or: "")
```
Comment (bugKrusha) [inline Sources/SlackBotKit/Camillink/Camillink+LinkMention.swift]: I think we can improve the clarity a bit here:
```
let recordString = try storage.get(key: linkURL.absoluteString, from: Keys.namespace, or: "")
guard !recordString.isEmpty, let recordData = recordString.data(using: .utf8) else { return nil }

let record = try JSONDecoder().decode(Record.self, from: recordData)
```

File: Sources/SlackBotKit/Camillink/Camillink+LinkMention.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        try bot.send(response.makeChatMessage())
+    }
+    
+    func markPreviousDiscussion(message: MessageDecorator, linkURL: URL    ) throws {
+        guard let channel = message.message.channel else { return }
+        // TODO: MAXG pending Chameleon change to get permalink
+//        storage.set(value: channel.id, forKey: linkURL.absoluteString, in: Keys.namespace)
+    }
+    
+    func isLinkWhitelisted(link: URL) throws -> Bool {
```
[the remark this answers is not in the document]
Comment (maxgoedjen) [reply Sources/SlackBotKit/Camillink/Camillink+LinkMention.swift]: Primarily in anticipation of future database access when implemented. Not needed for the moment now, I can remove.

File: Sources/SlackBotKit/Camillink/Camillink+Config.swift [hunk 4]
```diff
@@ -0,0 +1,19 @@
+
+extension CamillinkService {
+
```
[the remark this answers is not in the document]
Comment (maxgoedjen) [reply Sources/SlackBotKit/Camillink/Camillink+Config.swift]: Just the whitespace-only lines? Sure – is there any easy way to batch-do these/do we have a standard linter for this project?

File: Sources/SlackBotKit/Camillink/Camillink+LinkMention.swift [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        guard try !isLinkWhitelisted(link: linkURL) else { return }
+        if let previous = try previousLinkDiscussion(linkURL: linkURL) {
+            try sendPrompt(bot: bot, message: message, linkURL: linkURL, previousLink: previous)
+        } else {
+            try markPreviousDiscussion(message: message, linkURL: linkURL)
+        }
+
+    }
+    
+    func previousLinkDiscussion(linkURL: URL) throws -> URL? {
```
[the remark this answers is not in the document]
Comment (maxgoedjen) [reply Sources/SlackBotKit/Camillink/Camillink+LinkMention.swift]: Not terribly – it seemed like the convention was "throws is mostly for the API/db access exception-type handling" – ie, things that should probably cause the request to fail. This isn't so much that case – it still hits the DB and as such follows the `throw`ing convention, but the actual method itself seemed best expressed as an optional.

Open to suggestions though if my reading of this is off – this is the first Vapor project I've done more than play around with.

File: Sources/SlackBotKit/Camillink/Camillink+Config.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxgoedjen) [reply Sources/SlackBotKit/Camillink/Camillink+Config.swift]: 🤦‍♂ never mind had my Xcode settings on this laptop messed up. Fixing.

**Pull request merged**